hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/CompactStreamSerializerTest.java

```diff
@@ -101,6 +109,106 @@ public void testAllTypesWithCustomSerializer() {
+                .build();
+        EmptyDTO expected = new EmptyDTO();
+
+        Data data = serializationService.toData(expected);
+        EmptyDTO actual = serializationService.toObject(data);
+
+        assertEquals(expected, actual);
+    }
+
+    @Test
+    public void testVarSizedFieldsWithReflectiveSerializer() {
+        SerializationService serializationService = createSerializationService();
+        VarFieldsDTO expected = createVarFieldsDTO();
+
+        Data data = serializationService.toData(expected);
+        VarFieldsDTO actual = serializationService.toObject(data);
+
+        assertEquals(expected, actual);
+    }
+
+    @Test
+    public void testVarSizedFieldsWithCustomSerializer() {
+        CompactSerializationConfig compactSerializationConfig = new CompactSerializationConfig();
+        compactSerializationConfig.addSerializer(new VarFieldsDTOSerializer());
+        compactSerializationConfig.addSerializer(new InnerDTOSerializer());
+        SerializationService serializationService = new DefaultSerializationServiceBuilder()
+                .setSchemaService(schemaService)
+                .setConfig(new SerializationConfig().setCompactSerializationConfig(compactSerializationConfig))
+                .build();
+        VarFieldsDTO expected = createVarFieldsDTO();
+
+        Data data = serializationService.toData(expected);
+        VarFieldsDTO actual = serializationService.toObject(data);
+
+        assertEquals(expected, actual);
+    }
+
+    @Test
+    public void testFixedSizedFieldsWithReflectiveSerializer() {
+        SerializationService serializationService = createSerializationService();
+        FixedFieldsDTO expected = createFixedFieldsDTO();
+
+        Data data = serializationService.toData(expected);
+        FixedFieldsDTO actual = serializationService.toObject(data);
+
+        assertEquals(expected, actual);
+    }
+
+    @Test
+    public void testFixedSizedFieldsWithCustomSerializer() {
+        CompactSerializationConfig compactSerializationConfig = new CompactSerializationConfig();
+        compactSerializationConfig.addSerializer(new FixedFieldsDTOSerializer());
+        compactSerializationConfig.addSerializer(new InnerDTOSerializer());
+        SerializationService serializationService = new DefaultSerializationServiceBuilder()
+                .setSchemaService(schemaService)
+                .setConfig(new SerializationConfig().setCompactSerializationConfig(compactSerializationConfig))
+                .build();
+        FixedFieldsDTO expected = createFixedFieldsDTO();
+
+        Data data = serializationService.toData(expected);
+        FixedFieldsDTO actual = serializationService.toObject(data);
+
+        assertEquals(expected, actual);
+    }
+
+    @Test
+    public void testTypeMismatchThrowsException() {
+        CompactSerializationConfig compactSerializationConfig = new CompactSerializationConfig();
```

Comment:
To make the PR smaller, maybe use generic records while writing, and read with some other serializer we already have, and delete the TypeMismatchDTOSerializer